src/mlpack/methods/ann/layer/layer_types.hpp

```diff
@@ -0,0 +1,117 @@
+#include <mlpack/methods/ann/convolution_rules/naive_convolution.hpp>
+#include <mlpack/methods/ann/convolution_rules/fft_convolution.hpp>
+
+namespace mlpack {
+namespace ann {
+
+template<typename InputDataType, typename OutputDataType> class AddMerge;
+template<typename InputDataType, typename OutputDataType> class Concat;
+template<typename InputDataType, typename OutputDataType> class DropConnect;
+template<typename InputDataType, typename OutputDataType> class Glimpse;
+template<typename InputDataType, typename OutputDataType> class Linear;
+template<typename InputDataType, typename OutputDataType> class LinearNoBias;
+template<typename InputDataType, typename OutputDataType> class LSTM;
+template<typename InputDataType, typename OutputDataType> class Recurrent;
+template<typename InputDataType, typename OutputDataType> class Sequential;
+template<typename InputDataType, typename OutputDataType> class VRClassReward;
+
+template<
+    typename OutputLayerType,
+    typename InputDataType,
+    typename OutputDataType
+>
+class ConcatPerformance;
+
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename InputDataType,
+    typename OutputDataType
+>
+class Convolution;
+
+template<
+    typename InputDataType,
+    typename OutputDataType
+>
+class RecurrentAttention;
+
+using LayerTypes = boost::variant<
+    Add<arma::mat, arma::mat>*,
+    AddMerge<arma::mat, arma::mat>*,
+    BaseLayer<LogisticFunction, arma::mat, arma::mat>*,
+    BaseLayer<IdentityFunction, arma::mat, arma::mat>*,
+    BaseLayer<TanhFunction, arma::mat, arma::mat>*,
+    BaseLayer<RectifierFunction, arma::mat, arma::mat>*,
+    Concat<arma::mat, arma::mat>*,
+    ConcatPerformance<NegativeLogLikelihood<arma::mat, arma::mat>,
+                      arma::mat, arma::mat>*,
+    Constant<arma::mat, arma::mat>*,
+    Convolution<NaiveConvolution<ValidConvolution>,
+                NaiveConvolution<FullConvolution>,
+                NaiveConvolution<ValidConvolution>, arma::mat, arma::mat>*,
+    DropConnect<arma::mat, arma::mat>*,
+    Dropout<arma::mat, arma::mat>*,
+    Glimpse<arma::mat, arma::mat>*,
+    HardTanH<arma::mat, arma::mat>*,
+    Join<arma::mat, arma::mat>*,
+    LeakyReLU<arma::mat, arma::mat>*,
+    Linear<arma::mat, arma::mat>*,
+    LinearNoBias<arma::mat, arma::mat>*,
+    LogSoftMax<arma::mat, arma::mat>*,
+    Lookup<arma::mat, arma::mat>*,
+    LSTM<arma::mat, arma::mat>*,
+    MaxPooling<arma::mat, arma::mat>*,
+    MeanPooling<arma::mat, arma::mat>*,
+    MeanSquaredError<arma::mat, arma::mat>*,
+    MultiplyConstant<arma::mat, arma::mat>*,
+    NegativeLogLikelihood<arma::mat, arma::mat>*,
+    Recurrent<arma::mat, arma::mat>*,
+    RecurrentAttention<arma::mat, arma::mat>*,
+    ReinforceNormal<arma::mat, arma::mat>*,
+    Select<arma::mat, arma::mat>*,
+    Sequential<arma::mat, arma::mat>*,
+    VRClassReward<arma::mat, arma::mat>*
+>;
```

Comment:
Wouldn't you have to make `Forward()`, `Backward()`, and `Gradient()` virtual for inheriting from `Layer<>` to work?

Anyway the `LocalTypes` solution is nicer than what I proposed.

I don't see a way using boost::variant to collect the full list of types used by `Add()` at compile time.  Anything I can think of is pretty ugly.  But I don't think it's a huge problem to ask users to use a template parameter if they want to use their own custom layer type---we can just try to provide enough types that it's not a problem.  And maybe if we think of something better later, we can refactor.  What do you think?